File: src/main/python/JXTABLES/BaseHandler.py
import threading
import logging
import zmq


class BaseHandler(threading.Thread):
    """
    BaseHandler - A general base class for threads with common setup and cleanup functionality.

    This class provides a standardized way to handle thread initialization, resource cleanup,
    and logging for threads used in the XTABLES project.

    Author: Kobe Lei
    Version: 1.0
    """

    # Set up a logger for the class
    logger = logging.getLogger("BaseHandler")
    logging.basicConfig(level=logging.INFO)

    def __init__(self, thread_name: str, is_daemon: bool, socket: zmq.Socket):
        """
        Constructor to initialize the thread with a custom name, daemon status, and socket.

        :param thread_name: The name of the thread
        :param is_daemon: Whether the thread should run as a daemon
        :param socket: The zmq.Socket instance to manage
        """
        super().__init__(name=thread_name, daemon=is_daemon)
        self.socket = socket

    def clean_up(self):
        """
        Performs cleanup by closing the zmq.Socket.
        """
        if self.socket is not None:
            try:
                self.socket.close()
                self.logger.info("Socket closed successfully")
            except Exception as e:
                self.logger.warning(f"Exception during socket cleanup: {e}")

    def run(self):
        """
        Override this method in subclasses to define thread execution logic.
        """
        raise NotImplementedError("Subclasses must implement the run method")

    def handle_exception(self, exception: Exception):
        """
        Logs exceptions that occur during thread execution.

        :param exception: The exception that occurred
        """
        self.logger.error(f"Exception in thread {self.name}: {exception}")

    def interrupt(self):
        """
        Handles cleanup and logs exceptions when the thread is interrupted.
        """
        try:
            self.clean_up()
        except Exception as e:
            self.logger.warning(f"Exception during cleanup: {e}")
        super().join(0)  # In Python, `join(0)` can be used to exit the thread

File: src/main/java/org/kobe/xbot/JClient/BaseHandler.java
package org.kobe.xbot.JClient;

import org.kobe.xbot.Utilities.Logger.XTablesLogger;
import org.zeromq.ZMQ;

/**
 * BaseThread - A general base class for threads with common setup and cleanup functionality.
 * <p>
 * This class provides a standardized way to handle thread initialization, resource cleanup,
 * and logging for threads used in the XTABLES project.
 * <p>
 * Author: Kobe Lei
 * Version: 1.0
 */
public abstract class BaseHandler extends Thread {
    protected static final XTablesLogger logger = XTablesLogger.getLogger();
    protected ZMQ.Socket socket;

    /**
     * Constructor to initialize the thread with a custom name, daemon status, and socket.
     *
     * @param threadName The name of the thread
     * @param isDaemon   Whether the thread should run as a daemon
     * @param socket     The ZMQ.Socket instance to manage
     */
    public BaseHandler(String threadName, boolean isDaemon, ZMQ.Socket socket) {
        setName(threadName);
        setDaemon(isDaemon);
        this.socket = socket;
    }

    /**
     * Performs cleanup by closing the ZMQ.Socket.
     */
    protected void cleanUp() {
        if (socket != null) {
            socket.close();
            logger.info("Socket closed successfully");
        }
    }

    /**
     * Logs exceptions and handles cleanup when the thread is interrupted.
     */
    @Override
    public void interrupt() {
        try {
            cleanUp();
        } catch (Exception e) {
            logger.warning("Exception during cleanup: " + e.getMessage());
        }
        super.interrupt();
    }

    /**
     * Logs exceptions that occur during thread execution.
     */
    protected void handleException(Exception exception) {
        exception.printStackTrace();
        logger.severe("Exception in thread " + getName() + ": " + exception.getMessage());
    }
}

